src/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -282,262 +280,530 @@ internal static ulong ToUInt64(object value)
+            TryParse<TEnum>(value, false, out result);
 
-                    case ParseFailureKind.ArgumentWithParameter:
-                        return new ArgumentException(SR.Format(SR.GetResourceString(m_failureMessageID), m_failureMessageFormatArgument));
+        public static bool TryParse<TEnum>(string value, bool ignoreCase, out TEnum result) where TEnum : struct
+        {
+            if (!typeof(TEnum).IsEnum)
+            {
+                throw new ArgumentException(SR.Arg_MustBeEnum, nameof(TEnum));
+            }
 
-                    case ParseFailureKind.UnhandledException:
-                        return m_innerException;
+            ReadOnlySpan<char> valueSpan = value.AsSpan().TrimStart();
+            if (valueSpan.Length == 0)
+            {
+                result = default;
+                return false;
+            }
 
-                    default:
-                        Debug.Fail("Unknown EnumParseFailure: " + m_failure);
-                        return new ArgumentException(SR.Arg_EnumValueNotFound);
-                }
+            switch (Type.GetTypeCode(typeof(TEnum)))
+            {
+                case TypeCode.SByte: return TryParseSByteEnum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                case TypeCode.Byte: return TryParseByteEnum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                case TypeCode.Int16: return TryParseInt16Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                case TypeCode.UInt16: return TryParseUInt16Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                case TypeCode.Int32: return TryParseInt32Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                case TypeCode.UInt32: return TryParseUInt32Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                case TypeCode.Int64: return TryParseInt64Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                case TypeCode.UInt64: return TryParseUInt64Enum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
+                default: return TryParseRareEnum<TEnum>((RuntimeType)typeof(TEnum), value, valueSpan, ignoreCase, throwOnFailure: false, out result);
             }
         }
 
-        public static bool TryParse(Type enumType, string value, out object result)
+        public static object Parse(Type enumType, string value) =>
+            Parse(enumType, value, false);
+
+        public static object Parse(Type enumType, string value, bool ignoreCase)
```

Comment:
Is there a way to implement the Parse in terms of TryParse to avoid code duplication (with an added throwOnFailure argument passed in)? There is only one place where we return false, so we know what exception to throw, no?

Reply:
Done.